### PR TITLE
Add CUDA 11.3 + PyTorch 1.11.0 and 1.12.1

### PR DESCRIPTION
**Pull request by thiagocrepaldi**
Currently CI supports only CUDA 11.1, which is sufficient for PyTorch versions up to 1.10.
For PyTorch 1.11+ and nightly builds, the minimum CUDA version is 11.3.

This PR adds 3 new pipelines:
* Linux CUDA 11.3 which installs PyTorch, TorchVision, Detectron2 nightly builds.
* Linux CUDA 11.3 which installs PyTorch 1.11 and 1.12 stable release with matching TorchVision and Detectron2 builds.
* Linux CPU which installs PyTorch, TorchVision, Detectron2 nightly builds.
* Windows CPU which install PyTorch, TorchVision, Detectron2 nightly builds.

The new pipelines are important to test end-to-end ONNX export tests, in which PyTorch inference results are numerically compared with the ONNX Runtime's . A second benefit is help identifying ONNX export issues earlier in the development cycle.

ps: Nightly builds are only installed in the new pipelines (aka `windows_cpu_build_pytorch_master`, `linux_cuda113_tests_pytorch_master_python39`, `linux_cpu_tests_pytorch_master`. All existing pipelines are intact.

### CODE REVIEW

**Comment by thiagocrepaldi** [comment]
@FrancescoMandru these new pipelines should guard ONNX export more effciently.
By using PyTorch master, we can keep detectron2 and PyTorch on par in a matter of days instead of from release to release

See what you think

**Comment by thiagocrepaldi** [comment]
@ppwwyyxx how do you feel about this one?

**Comment by thiagocrepaldi** [comment]
> @ppwwyyxx how do you feel about this one?

gentle ping

**Comment by ppwwyyxx** [comment]
pytorch nightly gets broken from time to time in my experience, so I'm not sure if it's a good idea to test regularly on it.
It might be good to test it but report errors as warnings but last time I checked circleci cannot report warnings to github

**Comment by thiagocrepaldi** [comment]
> pytorch nightly gets broken from time to time in my experience, so I'm not sure if it's a good idea to test regularly on it. It might be good to test it but report errors as warnings but last time I checked circleci cannot report warnings to github

The instabilities comes from pyTorch tests itself, not the repo. Several tests are flaky, randomly fail and all.
However, my proposal is to use pytorch nightly and only run detectron2 tests.

We can disable the pipeline if I am proven wrong, but I have been using these tests locally with several nightly builds with no issue

**Comment by thiagocrepaldi** [comment]
Depends on https://github.com/facebookresearch/detectron2/pull/4120/

**Comment by thiagocrepaldi** [comment]
@ppwwyyxx due to the long time to have PRs merged, I have expanded the scope of this PR by combining three others.
Could you review this one, when you get a chance?

**Comment by thiagocrepaldi** [comment]
Trying to figure out this issue on the pipeline https://github.com/pytorch/pytorch/issues/81908

**Comment by thiagocrepaldi** [comment]
The `testKeypointHead` is currently failing as it depends on the [still-open PR](https://github.com/facebookresearch/detectron2/pull/4315)

**Comment by thiagocrepaldi** [comment]
@ppwwyyxx @mcimpoi Could you guys take a look at this PR when you have some time?

File: detectron2/modeling/meta_arch/retinanet.py [hunk 1]
```diff
@@ -353,11 +353,16 @@ def norm(c):
...
+            try:
+                norm_name = str(type(get_norm(norm, 1)))
+                if "BN" in norm_name:
+                    logger.warning(
+                        f"Shared BatchNorm (type={norm_name}) may not work well in RetinaNetHead."
+                    )
+            except ValueError:
+                # https://github.com/pytorch/pytorch/pull/74293
+                if isinstance(norm, str):
+                    logger.warning(f"No BatchNorm was found for '{norm}' for RetinaNetHead.")
```
Comment (ppwwyyxx) [inline detectron2/modeling/meta_arch/retinanet.py:365]: The correct fix is https://github.com/facebookresearch/detectron2/pull/4470

**Comment by thiagocrepaldi** [comment]
Contains https://github.com/facebookresearch/detectron2/pull/4470

**Comment by thiagocrepaldi** [comment]
@wat3rBro Could you review this one when you have some time? Thanks

**Comment by thiagocrepaldi** [comment]
> @ppwwyyxx @mcimpoi Could you guys take a look at this PR when you have some time?

gentle ping

**Comment by thiagocrepaldi** [comment]
Rebasing to allow up-to-date review

**Comment by wat3rBro** [comment]
@thiagocrepaldi sorry for the long long wait. I plan to do some holiday cleanings to bump up the PyTorch versions for D2 especially for CIs, this PR is a very good starting point, wondering do you still plan to work on this?

**Comment by thiagocrepaldi** [comment]
> @thiagocrepaldi sorry for the long long wait. I plan to do some holiday cleanings to bump up the PyTorch versions for D2 especially for CIs, this PR is a very good starting point, wondering do you still plan to work on this?

@wat3rBro I haven't worked on this for a while, but I can try to help

**Comment by wat3rBro** [comment]
@thiagocrepaldi Thanks for the update! The solution for installing CUDA 11.3 looks great, I was struggling making PyTorch 1.11+ works with stock ubuntu image in https://github.com/facebookresearch/detectron2/pull/4269, your PR solves the problem. Now the nightly requires CUDA 11.7, but I think we can install it in the similar way.

Moving forward I suggest maintaining 3 latest stable PyTorch versions + the nightly, right now it means:
- PyTorch 1.11 + CUDA 11.3
- PyTorch 1.12 + CUDA 11.3/11.6
- PyTorch 1.13 + CUDA 11.6/11.7
- PyTorch nightly + CUDA 11.7/11.8
- Drop the support for 1.08, 1.09 and 1.10
In this PR, we can first add PyTorch 1.11 and nightly, and later add 1.12&1.13 and remove old versions, does it sound good? cc. @ppwwyyxx

File: .circleci/config.yml [hunk 2]
```diff
@@ -115,15 +152,29 @@ install_linux_dep: &install_linux_dep
...
         pip install --progress-bar off ninja opencv-python-headless!=4.7.0.68 pytest tensorboard pycocotools onnx
-        pip install --progress-bar off torch==$PYTORCH_VERSION -f $PYTORCH_INDEX
-        if [[ "$TORCHVISION_VERSION" == "master" ]]; then
-          pip install git+https://github.com/pytorch/vision.git
+        if [[ "$PYTORCH_VERSION" == "master" ]]; then
+          echo "Installing torch/torchvision from $PYTORCH_INDEX"
+          # Remove first, in case it's in the CI cache
+          pip uninstall -y torch torchvision
+          pip install -v --progress-bar off --pre torch torchvision --extra-index-url $PYTORCH_INDEX
+          pip install -v --progress-bar off omegaconf==2.1.2   # needed by FCOSE2ETest::test_empty_data
```
Comment (ppwwyyxx) [inline .circleci/config.yml:160]:
```suggestion
          pip install -v --progress-bar off omegaconf==2.1.2
```

It's fine to lock the version, but I don't think it's required after https://github.com/facebookresearch/detectron2/pull/4253

Comment (thiagocrepaldi) [reply .circleci/config.yml]: removed the version pinning for omegaconf

**Comment by thiagocrepaldi** [comment]
@wat3rBro is something like this that you were thinking?

**Comment by wat3rBro** [comment]
@thiagocrepaldi Thanks for updating it, overall it looks promising. Some of the CIs didn't pass because indexes mismatch, have you found the reason?

File: .circleci/config.yml [hunk 3]
```diff
@@ -208,7 +284,8 @@ jobs:
           paths:
             - /opt/circleci/.pyenv
             - ~/.torch
-          key: cache-{{ arch }}-<< parameters.pytorch_version >>-{{ .Branch }}-20210827
+          key: cache-{{ arch }}-<< parameters.pytorch_version >>-{{ .Branch }}-20990004
+
```
Comment (wat3rBro) [inline .circleci/config.yml:288]: There're a lot of duplication between `linux_cuda111_tests`, `linux_cuda113_tests` and `linux_cuda117_tests`, it would be nice to make it a bit more concise.

**Comment by thiagocrepaldi** [comment]
> @thiagocrepaldi Thanks for updating it, overall it looks promising. Some of the CIs didn't pass because indexes mismatch, have you found the reason, it seems that it changes how `pytorch` and `torchvision` are installed.

yeah, i saw that and I honestly ran out of tricks to force the right cuda version. It seems that by PEP400 (if I am not mistaken), the "+<string>" is called "local version label" and it is not used for version selection. The pytorch's URL for cu113 also has wheels for a bunch of other cuda versions, and this mismatch is happening

File: .circleci/config.yml [hunk 4]
```diff
@@ -115,15 +151,29 @@ install_linux_dep: &install_linux_dep
...
+        python -c 'import torchvision; print("TorchVision Version:", torchvision.__version__)'
         python -c 'import torch; print("CUDA:", torch.cuda.is_available())'
+        echo "Python packages"
+        python -c "import sys; print(sys.executable)"
+        python --version
+        pip list
+        echo "GCC Compiler"
         gcc --version
+        echo "OS Environment Variables"
+        env
```
Comment (ppwwyyxx) [inline .circleci/config.yml:175]: There are usually secrets in envvars that should not be printed.
CircleCI has built something to prevent it: https://circleci.com/blog/keep-environment-variables-private-with-secret-masking but maybe better to just not do it.

File: .circleci/config.yml [hunk 5]
```diff
@@ -8,22 +8,27 @@ cpu: &cpu
...
   resource_class: medium
 
-gpu: &gpu
+cuda111: &cuda111
   machine:
     # NOTE: use a cuda version that's supported by all our pytorch versions
     image: ubuntu-1604-cuda-11.1:202012-01
   resource_class: gpu.nvidia.small
 
+cuda113: &cuda113
```
[the remark this answers is not in the document]
Comment (thiagocrepaldi) [reply .circleci/config.yml:17]: How would that work?

For cuda 11.3, we uninstall cuda 11.4 and install 11.3
For cuda 11.7, we uninstall cuda 11.4 and install 11.7

cuda 11.4 is a common factor between 11.3 and 11.7, but each cuda needs to extend the 11.4 base

File: .circleci/config.yml (hunk 3 above)
Comment (thiagocrepaldi) [reply .circleci/config.yml:288]: I am not really familiar with CircleCI. I've just winged based on sampels. Any ideas on how to templatize even more?
The `- <<:` construction is already an abstraction layer, but maybe we can group them in a larger template if CircleCI has such mechanism

File: .circleci/config.yml [hunk 6]
```diff
@@ -249,23 +342,60 @@ workflows:
...
+          name: linux_cuda111_tests_pytorch1.9
           pytorch_version: '1.9+cu111'
           torchvision_version: '0.10+cu111'
-      - linux_gpu_tests:
-          name: linux_gpu_tests_pytorch1.10
-          pytorch_version: '1.10+cu111'
+      - linux_cuda111_tests:
+          name: linux_cuda111_tests_pytorch1.10
+          pytorch_version: '1.10.0+cu111'
           torchvision_version: '0.11.1+cu111'
```
[the remark this answers is not in the document]
Comment (thiagocrepaldi) [reply .circleci/config.yml:368]: I was thinking in having a PR just to remove the old stuff. I want to make sure the new one is solid before removing something that is reliable already

File: .circleci/config.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (thiagocrepaldi) [reply .circleci/config.yml]: got it, the dupe was on the other PR lol
agreed, renamed to cuda114

The cuda 11.7 PR will remain kind of stale until this one is done. Then I can rebase that one on top of this and just add the required delta

File: .circleci/config.yml [hunk 7]
```diff
@@ -196,8 +267,10 @@ jobs:
 
       - restore_cache:
           keys:
-            - cache-{{ arch }}-<< parameters.pytorch_version >>-{{ .Branch }}-20210827
+            - cache-{{ arch }}-<< parameters.pytorch_version >>-{{ .Branch }}-20990005
 
+      - <<: *setupcuda113
+      - <<: *removecuda114
```
Comment (wat3rBro) [inline .circleci/config.yml:273]: shouldn't it remove 114 first then setup 113?

Comment (thiagocrepaldi) [reply .circleci/config.yml:273]: Removing 11.4 first and installing 11.3 would result in uninstalling several shared components only to reinstall later, increasing installation time. Keeping the current order does not affect the end result and save some time

**Comment by thiagocrepaldi** [comment]
@wat3rBro everything is green now

**Comment by thiagocrepaldi** [comment]
@wat3rBro gentle ping

**Comment by thiagocrepaldi** [comment]
@wat3rBro @ppwwyyxx gentle ping

**Comment by thiagocrepaldi** [comment]
Is this still relevant or should we close it?

**Review by kaustubhsapru — APPROVED** [review]
